MNT accelerate example plot_lle_digits.py

siavrez opened the pull request:
Changing matplotlib.text with matplotlib.scatter and Markers created from TeX symbols improves runtime by almost 25 seconds.
Also changed time.time with time.perf_counter.

<!--
Thanks for contributing a pull request! Please ensure you have taken a look at
the contribution guidelines: https://github.com/scikit-learn/scikit-learn/blob/main/CONTRIBUTING.md
-->

#### Reference Issues/PRs
https://github.com/scikit-learn/scikit-learn/issues/21598


#### What does this implement/fix? Explain your changes.


#### Any other comments?


<!--
Please be aware that we are a loose team of volunteers so patience is
necessary; assistance handling other issues is very welcome. We value
all user contributions, no matter how minor they are. If we are slow to
review, either the pull request needs some benchmarking, tinkering,
convincing, etc. or more likely the reviewers are simply busy. In either
case, we ask for your understanding during the review process.
For more information, see our FAQ on this topic:
http://scikit-learn.org/dev/faq.html#why-is-my-pull-request-not-getting-any-attention.

Thanks for contributing!
-->

siavrez committed:
accelerate examlpe ../examples/manifold/plot_lle_digits.py by changing plt.text with plt.scatter with Markers created from TeX symbols

glemaitre commented on examples/manifold/plot_lle_digits.py:
```diff
@@ -51,17 +51,13 @@
 def plot_embedding(X, title, ax):
     X = MinMaxScaler().fit_transform(X)
 
+    for t in np.unique(y):
```
I would propose the same renaming than in the previous example that I reviewed.

glemaitre commented on examples/manifold/plot_lle_digits.py:
```diff
@@ -151,7 +147,7 @@ def plot_embedding(X, title, ax):
 # of the original data. We will store the projected data as well as the computational
 # time needed to perform each projection.
-from time import time
+from time import perf_counter
```
I would also keep `time` here.

glemaitre reviewed (commented):
I don't know if there is a way to accelerate even more the example by tweaking MDS, t-SNE, and NCA to accelerate them.

siavrez committed:
accelerate plot_lle_digits.py changed proposed namings and params for TSNE, NCA, MDS

siavrez commented:
Added n_iter=500, n_iter_without_progress=150, n_jobs=-1 to TSNE , n_jobs=-1 to MDS and changed init="random" to init="pca" in NCA.
Runtime is now 42 p/m 6 seconds.
from 72 p/m 12 seconds.

jjerphan commented on examples/manifold/plot_lle_digits.py:
```diff
@@ -51,17 +51,17 @@
 def plot_embedding(X, title, ax):
     X = MinMaxScaler().fit_transform(X)
 
+    for digit in digits.target_names:
+        ax.scatter(
+            *X[y == digit].T,
+            marker=f"${digit}$",
+            s=60,
+            color=plt.cm.Dark2(digit),
+            alpha=0.7,
+        )
```
Does this addition (and the deletion bellow) change the rendering of this example?

siavrez replied:
Default plot
![previous](https://user-images.githubusercontent.com/25286805/143011874-66252390-e873-483e-923c-0edaac76a0e8.png)

after this changes
![newlook](https://user-images.githubusercontent.com/25286805/143012002-8bf55d9c-19c8-49ca-8fb5-005429bd3db8.png)

changing zorder:
![changedZorder](https://user-images.githubusercontent.com/25286805/143012067-4eea3ecf-4565-4c63-8fe9-4c13d3fdfbb4.png)

jjerphan commented on examples/manifold/plot_lle_digits.py:
```diff
@@ -131,7 +131,7 @@ def plot_embedding(X, title, ax):
         n_neighbors=n_neighbors, n_components=2, method="ltsa"
     ),
-    "MDS embedding": MDS(n_components=2, n_init=1, max_iter=100),
+    "MDS embedding": MDS(n_components=2, n_init=1, max_iter=100, n_jobs=-1),
```
Have you tried changing `max_iter`, here?

siavrez replied:
`max_iter ` bellow 100 is not converging  and there's room to go to 125.
`n_jobs ` is helping reduce runtime for MDS. for n_jobs=2 it is about 0.25 percent.
with `max_iter `=125 and `n_jobs`=2 it takes 3.9 +/- 0.2  seconds.

jjerphan reviewed (commented):
Thank you for tackling this, @siavrez.

Here are two comments.

glemaitre commented:
Let's use only `n_jobs=2` instead of `n_jobs=-1`. It would be nicer with your user.

siavrez committed:
Added zorder for annotationBox and scatter, changed n_jobs=2 and MDS max_iter=125

siavrez committed:
removed zorder in Annotationbox and set in after creating object

siavrez commented:
Previous commit failed because of a signature test in matplotlib. So removed the param (zorder) from the initial call and set it later.

adrinjalali reviewed (approved):
Thanks @siavrez

jjerphan commented:
Thank you, @siavrez, for this PR. Having plots being reported is helpful for contributors. :+1:

jjerphan reviewed (approved):
Thank you, @siavrez, for this PR. Having plots being reported is helpful for reviewers. +1
